### ISSUE TITLE
serializing SDO with custom properties fails

### ISSUE THREAD

**Issue by tlhampton13** [oasis-open/cti-python-stix2#179]
I am using custom properties on the domain objects, but the serialize method fails if the SDO has any custom properties.  

```
sdo = stix2.AttackPattern(
        name='AttackName',
        custom_properties={
            'x_score': 5
        }
    )
json = sdo.serialize()
print(json)
```

produces this error:

```
stix2.exceptions.ExtraPropertiesError: Unexpected properties for AttackPattern: (x_score)
```

**Comment by clenk** [comment]
Hi @tlhampton13! We just pushed a fix (https://github.com/oasis-open/cti-python-stix2/pull/178) for a similar bug with custom properties. Do you still see the error with the latest code on the master branch?

**Comment by tlhampton13** [comment]
@clenk No, it still fails.  Looks like the copy.deepcopy() function is where it breaks.

**Comment by tlhampton13** [comment]
@clenk It does work if I use either of these approaches

```
    identity = stix2.Identity(name="John Smith",
                              identity_class="individual",
                              allow_custom=True,
                              custom_properties={"x_foo": "bar"})
    json = identity.serialize()
    print(json)
```

```
    identity = stix2.Identity(name="John Smith",
                              identity_class="individual",
                              allow_custom=True,
                              x_foo="bar")
    json = identity.serialize()
    print(json)
```

I would not expect to need to set allow_custom=True in the first example, because I'm supplying the custom_properties dict as an argument.  I would think the class could figure out to auto set allow_custom=True when the custom_properties argument is provided.

When I just provide the custom_properties argument and do NOT set allow_custom=True to it fails in base.__deepcopy__ line 212 when constructing the new class from a dictionary, because allow_custom was never set.

I guess, a solution would be to set allow_custom=True in the base.__init__ when the custom_properties argument is provided.  Although, I'm not appreciating why there is two different ways to pass custom properties.

**Comment by clenk** [comment]
Thanks for the additional info, @tlhampton13! Let me know if the latest code (#180) doesn't fix the issue for you.
